Flashcard console app ready for code review

ngalantino opened the pull request.

lawangGurung commented on Flashcards.ngalantino/Flashcards.ngalantino/StudySession.cs:
```diff
@@ -0,0 +1,5 @@
+public class StudySession {
+    public string Stack { get; set; }
```
string in c# is a non-nullable type, so try to make the string property nullable or initialize the value.
"public string? Stack {get; set;}" or
"public string Stack {get; set;} = '' "

lawangGurung commented on Flashcards.ngalantino/Flashcards.ngalantino/Stack.cs:
```diff
@@ -0,0 +1,4 @@
+public class Stack {
+    public string Name { get; set; }
```
Try to make a string property nullable or assign an empty string as a value.

lawangGurung commented:
Your application doesn't ensure the database is created automatically,
please make the changes in your code base where the table for flashcards is created when the user runs the application.
(Hint: Always make sure the flashcard table is created when there is no flashcard table in the database).

lawangGurung reviewed (changes requested):
Good luck!

ngalantino commented:
Thanks for reviewing my code :) 

I added some code in the constructor of DatabaseManager to create the flaschard table if it does not exist. Also made string properties nullable in Stack and StudySession.

lawangGurung commented:
![Screenshot from 2024-11-06 21-29-19](https://github.com/user-attachments/assets/1ac4fccf-f32c-4f2f-8f17-158332fb1ff2)
- Fix this error
- You have only ensured the table is created automatically not the DATABASE. but if you don't wanna do it add a readme file to suggest the user or reviewer of creating the database manually.
- You can seed some custom data for the testing purposes.

lawangGurung commented on Flashcards.ngalantino/Flashcards.ngalantino/StudyContentController.cs:
```diff
@@ -0,0 +1,191 @@
+        Console.WriteLine("Enter Id of flashcard to be updated: ");
+        int id = Int32.Parse(Console.ReadLine()) - 1;
+
+        flashcard.Id = flashcards[id].Id;
```
what if user enters the null value?

lawangGurung commented on Flashcards.ngalantino/Flashcards.ngalantino/StudyContentController.cs:
```diff
@@ -0,0 +1,191 @@
+
+        Console.WriteLine("Enter back of flashcard.");
+
+        flashcard.Back = Console.ReadLine();
```
What if user just presses 'ENTER'

lawangGurung reviewed (changes requested):
I have suggested some changes in the comment section please see through it, delete your flashcards database, and create a new flashcards database to see the errors.
Please test your code thoroughly and test the edge cases properly before submitting pr.

ngalantino commented:
@lawangGurung Thank you for the feedback! I have made the following changes: 

- Deleted database and implemented code to create database and tables if they do not exist.
- Added input validation for edge cases.

The database connection string can be configured in the App.config file.
<add key="dbConnection" value="data source=<**your_local_machine**>; trusted_connection=true" />

lawangGurung commented:
![Screenshot from 2024-11-09 12-52-53](https://github.com/user-attachments/assets/164abbde-143d-4840-8753-fef53a7d55a2)

- When exiting the study area an exception is thrown. Please see to it.
![Screenshot from 2024-11-09 12-52-53](https://github.com/user-attachments/assets/7112c7d5-c0f4-41d2-9cab-94124b81ed12)
- when giving the non-integer value to the 'Enter Id' in both update and delete will also throw exception. You can validate the input using 'AnsiConsole.Ask<int>()'.

lawangGurung reviewed (changes requested):
Please the changes are applied.

ngalantino commented:
I added try/catch blocks to validate input for updating and deleting flashcards. I wasn't able to duplicate the exception when exiting the study session area. "Conversion failed when converting date and/or time from character string". What could cause that exception?

lawangGurung commented on Flashcards.ngalantino/Flashcards.ngalantino/DatabaseManager.cs:
```diff
@@ -0,0 +1,270 @@
+        }
+    }
+
+    public void AddStudySession(String stack, DateTime date, int score)
+    {
+
+        using (SqlConnection connection = new SqlConnection(System.Configuration.ConfigurationManager.AppSettings.Get("dbConnection")))
+        {
+
+            connection.Open();
+
+            string sql = $"INSERT INTO flashcardstest.dbo.studysessions (stack, date, score) VALUES ('{stack}', '{date}', '{score}')";
+
+            SqlCommand command = new SqlCommand(sql, connection);
```
You have an error in this SQL statement. Use parameters to avoid further errors. If in doubt, refer to the Dapper documentation.

lawangGurung reviewed (changes requested):
Read the error you are getting and check the SQL statement, USE PARAMETER.
